native/com_wolfssl_WolfSSL.c

```diff
@@ -678,6 +686,138 @@ void NativeLoggingCallback(const int logLevel, const char *const logMessage)
+    if ((*jenv)->ExceptionOccurred(jenv)) {
+        (*jenv)->ExceptionDescribe(jenv);
+        (*jenv)->ExceptionClear(jenv);
+        return;
+    }
+
+    /* check if our stored object reference is valid */
+    refcheck = (*jenv)->GetObjectRefType(jenv, g_fipsCbIfaceObj);
+    if (refcheck == 2) {
+
+        /* lookup WolfSSLLoggingCallback class from global object ref */
+        jclass fipsCbClass = (*jenv)->GetObjectClass(jenv, g_fipsCbIfaceObj);
+        if (!fipsCbClass) {
+            if ((*jenv)->ExceptionOccurred(jenv)) {
+                (*jenv)->ExceptionDescribe(jenv);
+                (*jenv)->ExceptionClear(jenv);
+            }
+
+            (*jenv)->ThrowNew(jenv, excClass,
+                "Can't get native WolfSSLFIPSErrorCallback class reference");
+            return;
+        }
+
+        errorMethod = (*jenv)->GetMethodID(jenv, fipsCbClass,
+                                            "errorCallback",
+                                            "(IILjava/lang/String;)V");
+        if (errorMethod == 0) {
+            if ((*jenv)->ExceptionOccurred(jenv)) {
+                (*jenv)->ExceptionDescribe(jenv);
+                (*jenv)->ExceptionClear(jenv);
+            }
+            (*jenv)->ThrowNew(jenv, excClass,
+                "Error getting errorCallback method from JNI");
+            return;
+        }
+
+        /* create jstring from char* */
+        jstring hashString = (*jenv)->NewStringUTF(jenv, hash);
+
+        (*jenv)->CallVoidMethod(jenv, g_fipsCbIfaceObj, errorMethod,
+                ok, err, hashString);
+
+        if ((*jenv)->ExceptionOccurred(jenv)) {
+            (*jenv)->ExceptionDescribe(jenv);
+            (*jenv)->ExceptionClear(jenv);
+
+            (*jenv)->ThrowNew(jenv, excClass,
+                    "Error calling FIPS error callback from JNI");
+            return;
+        }
+
+    } else {
+        if ((*jenv)->ExceptionOccurred(jenv)) {
+            (*jenv)->ExceptionDescribe(jenv);
+            (*jenv)->ExceptionClear(jenv);
+        }
+
+        (*jenv)->ThrowNew(jenv, excClass,
+                "Object reference invalid in NativeFIPSErrorCallback");
+    }
+#endif
+}
+
+JNIEXPORT jint JNICALL Java_com_wolfssl_WolfSSL_setFIPSCb
+  (JNIEnv* jenv, jclass jcl, jobject callback)
+{
+    int ret = SSL_SUCCESS;
+    (void)jcl;
+
+#ifdef HAVE_FIPS
+    if (jenv == NULL || callback == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    /* store Java FIPS callback Interface object */
+    g_fipsCbIfaceObj = (*jenv)->NewGlobalRef(jenv, callback);
+    if (!g_fipsCbIfaceObj) {
+        printf("error storing global wolfCrypt FIP callback interface\n");
+        return SSL_FAILURE;
+    }
+
+    /* register NativeFIPSErrorCallback, wraps Java callback */
+    ret = wolfCrypt_SetCb_fips(NativeFIPSErrorCallback);
+    if (ret == 0) {
+        ret = SSL_SUCCESS;
+    }
+#else
+    (void)jenv;
+    (void)callback;
+    /* simply return SSL_SUCCESS if not compiled in / not FIPS lib */
```

Comment:
if fips is attempted to be used in a non fips build we should throw an error or reference back to wolfSSL to get fips

Reply:
Sounds good, adjusted setFIPSCb() to return NOT_COMPILED_IN when HAVE_FIPS is not defined.